<a href="https://colab.research.google.com/github/DevilNReality/Capstone-1-Used-Car-Price-Prediction/blob/main/Code%20File/%20Capstone_1_Used_Car_Prices.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Author Details:**

Programmed By: Karthikeyan V

GitHub: DevilNReality

Version: 1.0


# Approach To The Problem


1. Import Libraries And Datasets
2. Exploratory Data Analysis
3. Data Preparation
4. Train Test Split
5. Models
  1. Linear Regression
  2. Decision Tree Regressor
  3. Random Forest Regressor
  4. AdaBoosting Regressor
  5. GradientBoosting Regressor
  6. XG Boost Regressor
  7. Support Vector Regressor

# Install Libraries And Dependencies

In [ ]:
! pip install dataprep

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 152.6/152.6 kB 8.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.4/6.4 MB 70.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
INFO: pip is looking at multiple versions of dask[array,dataframe,delayed] to determine which version is compatible with other requirements. This could take a while.
INFO: pip is still looking at multiple versions of dask[array,dataframe,delayed] to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of dask-expr to determine which version is compatible with other requirements. This could take a whi

In [ ]:
! pip install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 364.4/364.4 kB 8.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.5/233.5 kB 15.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 59.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 5.4 MB/s eta 0:00:00
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 1.3.24
    Uninstalling SQLAlchemy-1.3.24:
      Successfully uninstalled SQLAlchemy-1.3.24
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
dataprep 0.4.5 requires sqlalchemy==1.3.24, but you have sqlalchemy 2.0.36 which is incompatible.
langchain 0.3.7 requires pydantic<3.0.0,>=2.7.4, but you have pydantic 1.10.19 which is incompatible.


In [ ]:
! pip install gradio

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 149.4/149.4 kB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.1/57.1 MB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 320.1/320.1 kB 12.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.9/94.9 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 434.9/434.9 kB 24.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.1/11.1 MB 49.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.2/73.2 kB 4.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.7/63.7 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 130.2/130.2 kB 4.5 MB/s eta 0:00:00
  Attempting uninstall: markupsafe
    Found existing installation: MarkupSafe 3.0.2
    Uninstalling MarkupSafe-3.0.2:
      Successfully uninstalled MarkupSafe-3.0.2
  Attempting uninstall: pydantic
    Found existing installation: pydantic 1.10.19
    Uninstalling pydantic-1.10.19:

# Load The Libaries and Dataset

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import optuna
import gradio as gr

from sklearn.model_selection import train_test_split , cross_val_score
from sklearn.preprocessing import LabelEncoder

from sklearn.metrics import r2_score , mean_squared_error
from sklearn.neighbors import KNeighborsRegressor
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor , AdaBoostRegressor , GradientBoostingRegressor
from xgboost import XGBRegressor
from sklearn.svm import SVR

from dataprep.eda import create_report
import re
import warnings
warnings.filterwarnings('ignore')

In [ ]:
! unzip '/content/Used Car Prices.zip'

Archive:  /content/Used Car Prices.zip
  inflating: sample_submission.csv   
  inflating: test.csv                
  inflating: train.csv               


In [ ]:
def load_data(path):
  '''
  Loads the data from given path

  Args:
    path : Path of the file

  Return:
    Dataframe of the given path
  '''
  data = pd.read_csv(path)
  return data

file_path = '/content/train.csv'
dataset = load_data(file_path)

## Audit the Dataset

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188533 entries, 0 to 188532
Data columns (total 13 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   id            188533 non-null  int64 
 1   brand         188533 non-null  object
 2   model         188533 non-null  object
 3   model_year    188533 non-null  int64 
 4   milage        188533 non-null  int64 
 5   fuel_type     183450 non-null  object
 6   engine        188533 non-null  object
 7   transmission  188533 non-null  object
 8   ext_col       188533 non-null  object
 9   int_col       188533 non-null  object
 10  accident      186081 non-null  object
 11  clean_title   167114 non-null  object
 12  price         188533 non-null  int64 
dtypes: int64(4), object(9)
memory usage: 18.7+ MB


In [ ]:
dataset.head()

id          brand              model  model_year  milage      fuel_type   
0   0           MINI      Cooper S Base        2007  213000       Gasoline  \
1   1        Lincoln              LS V8        2002  143250       Gasoline   
2   2      Chevrolet  Silverado 2500 LT        2002  136731  E85 Flex Fuel   
3   3        Genesis   G90 5.0 Ultimate        2017   19500       Gasoline   
4   4  Mercedes-Benz        Metris Base        2021    7388       Gasoline   

                                              engine   
0       172.0HP 1.6L 4 Cylinder Engine Gasoline Fuel  \
1       252.0HP 3.9L 8 Cylinder Engine Gasoline Fuel   
2  320.0HP 5.3L 8 Cylinder Engine Flex Fuel Capab...   
3       420.0HP 5.0L 8 Cylinder Engine Gasoline Fuel   
4       208.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   

                     transmission ext_col int_col   
0                             A/T  Yellow    Gray  \
1                             A/T  Silver   Beige   
2                             A/T    Blue    Gray   
3  Transmission w/Dual Shift Mode   Black   Black   
4                     7-Speed A/T   Black   Beige   

                                 accident clean_title  price  
0                           None reported         Yes   4200  
1  At least 1 accident or damage reported         Yes   4999  
2                           None reported         Yes  13900  
3                           None reported         Yes  45000  
4                           None reported         Yes  97500

Things found after auditing the data:
* It has 188533 rows of data.
* It has 12 columns of data
* Three columns have missing data

# Exploratory Data Analysis

In [ ]:
create_report(dataset)

EDA gave us:
1. Under Overview
  * Clean Title has highest missing value of 11 percent.
  * Model and Engine have high cardinality.
2. Under Variables
  * The need of cleaning Engine,Transmission,Ext_col,Int_col,Fuel Type columns.
  * Filling the Clean Title with "NO".

# Data Preparation

## Transforming Data

In [ ]:
# Getting Unique transmission data
dataset["transmission"].unique()

array(['A/T', 'Transmission w/Dual Shift Mode', '7-Speed A/T',
       '8-Speed A/T', '10-Speed Automatic', '1-Speed A/T', '6-Speed A/T',
       '10-Speed A/T', '9-Speed A/T', '8-Speed Automatic',
       '9-Speed Automatic', '5-Speed A/T', 'Automatic',
       '7-Speed Automatic with Auto-Shift', 'CVT Transmission',
       '5-Speed M/T', 'M/T', '6-Speed M/T', '6-Speed Automatic',
       '4-Speed Automatic', '7-Speed M/T', '2-Speed A/T',
       '1-Speed Automatic', 'Automatic CVT', '4-Speed A/T',
       '6-Speed Manual', 'Transmission Overdrive Switch',
       '8-Speed Automatic with Auto-Shift', '7-Speed Manual',
       '7-Speed Automatic', '9-Speed Automatic with Auto-Shift',
       '6-Speed Automatic with Auto-Shift',
       '6-Speed Electronically Controlled Automatic with O', 'F', 'CVT-F',
       '8-Speed Manual', 'Manual', '–', '2', '6 Speed At/Mt',
       '5-Speed Automatic', '2-Speed Automatic', '8-SPEED A/T', '7-Speed',
       'Variable', 'Single-Speed Fixed Gear', '8-SPEED AT',


In [ ]:
transmission_mapping = {
    '6-Speed A/T': 'Automatic',
    '8-Speed Automatic': 'Automatic',
    'Automatic': 'Automatic',
    '7-Speed A/T': 'Automatic',
    'A/T': 'Automatic',
    '8-Speed A/T': 'Automatic',
    'Transmission w/Dual Shift Mode': 'Automatic',
    '9-Speed Automatic': 'Automatic',
    '10-Speed Automatic': 'Automatic',
    'Automatic CVT': 'CVT',
    'CVT Transmission': 'CVT',
    'CVT-F': 'CVT',
    '7-Speed Automatic': 'Automatic',
    '6-Speed M/T': 'Manual',
    '7-Speed M/T': 'Manual',
    '6-Speed Manual': 'Manual',
    '5-Speed M/T': 'Manual',
    'Manual': 'Manual',
    '7-Speed Manual': 'Manual',
    '8-Speed Manual': 'Manual',
    'M/T': 'Manual',
    '6-Speed Automatic with Auto-Shift': 'Automatic',
    '8-Speed Automatic with Auto-Shift': 'Automatic',
    '9-Speed Automatic with Auto-Shift': 'Automatic',
    '5-Speed Automatic': 'Automatic',
    '4-Speed A/T': 'Automatic',
    '4-Speed Automatic': 'Automatic',
    '10-Speed A/T': 'Automatic',
    '5-Speed A/T': 'Automatic',
    '1-Speed A/T': 'Automatic',
    '2-Speed Automatic': 'Automatic',
    '6-Speed Electronically Controlled Automatic with O': 'Automatic',
    'Single-Speed Fixed Gear': 'Automatic',
    '7-Speed DCT Automatic': 'Automatic',
    '8-SPEED AT': 'Automatic',
    '8-SPEED A/T': 'Automatic',
    '10-Speed Automatic with Overdrive': 'Automatic',
    'Automatic, 9-Spd 9G-Tronic': 'Automatic',
    'Automatic, 8-Spd': 'Automatic',
    'Automatic, 8-Spd Sport w/Sport & Manual Modes': 'Automatic',
    'Automatic, 8-Spd PDK Dual-Clutch': 'Automatic',
    'Automatic, 8-Spd M STEPTRONIC w/Drivelogic, Sport & Manual Modes': 'Automatic',
    'Automatic, 8-Spd Dual-Clutch': 'Automatic',
    '2-Speed A/T': 'Automatic',
    '6 Speed At/Mt': 'Manual',
    '6 Speed Mt': 'Manual',
    'Variable': 'Other',
    '2': 'Other',
    '–': 'Other',
    'SCHEDULED FOR OR IN PRODUCTION': 'Other'}

dataset['transmission'] = dataset['transmission'].replace(transmission_mapping)
dataset['transmission'].value_counts()

transmission
Automatic                            160555
Manual                                16415
9-Speed A/T                            3866
CVT                                    2834
6-Speed Automatic                      2799
7-Speed Automatic with Auto-Shift      1172
1-Speed Automatic                       365
Other                                   250
Transmission Overdrive Switch           156
F                                        65
7-Speed                                  29
6-Speed                                  27
Name: count, dtype: int64

In [ ]:
def simplify_transmission(transmission):
    if 'Automatic' in transmission:
        return 'Automatic'
    elif 'Manual' in transmission:
        return 'Manual'
    else:
        return 'Other'

dataset['transmission'] = dataset['transmission'].apply(simplify_transmission)
dataset['transmission'].value_counts()

transmission
Automatic    164891
Manual        16415
Other          7227
Name: count, dtype: int64

**Mapping and transforming into the values - Automatic, Manual , Other.**

In [ ]:
dataset["engine"].unique()

array(['172.0HP 1.6L 4 Cylinder Engine Gasoline Fuel',
       '252.0HP 3.9L 8 Cylinder Engine Gasoline Fuel',
       '320.0HP 5.3L 8 Cylinder Engine Flex Fuel Capability', ...,
       '78.0HP 1.2L 3 Cylinder Engine Gasoline Fuel',
       '139.0HP 1.6L 4 Cylinder Engine Plug-In Electric/Gas',
       '313.0HP 2.0L 4 Cylinder Engine Plug-In Electric/Gas'],
      dtype=object)

In [ ]:
def extract_engine_attributes(engine_str):
    horsepower = re.search(r'(\d+\.\d+)HP|\d+\.\d+', engine_str)
    displacement = re.search(r'(\d+\.\d+L|\d+\.\d+ Liter)', engine_str)
    return horsepower.group(1) if horsepower else '',\
           displacement.group(1) if displacement else ''


dataset[['Horsepower', 'Engine_Displacement']] = dataset['engine'].apply(extract_engine_attributes).apply(pd.Series)
dataset['Horsepower'] = pd.to_numeric(dataset['Horsepower'], errors = 'coerce')
dataset['Engine_Displacement'] = dataset['Engine_Displacement'].str.replace('L', '')
dataset['Engine_Displacement'] = pd.to_numeric(dataset['Engine_Displacement'], errors = 'coerce')

In [ ]:
dataset["int_col"].unique()

array(['Gray', 'Beige', 'Black', '–', 'Blue', 'White', 'Red', 'Brown',
       'Dark Galvanized', 'Parchment.', 'Boulder', 'Orange',
       'Medium Earth Gray', 'Ebony', 'Canberra Beige', 'Jet Black',
       'Silver', 'Light Platinum / Jet Black', 'Macchiato/Magmagrey',
       'Gold', 'Cloud', 'Rioja Red', 'Global Black', 'Green',
       'Medium Stone', 'Navy Pier', 'Dark Ash', 'BLACK', 'Portland',
       'Sandstone', 'Canberra Beige/Black', 'Diesel Gray / Black',
       'Sarder Brown', 'Black Onyx', 'White / Brown', 'Black/Gun Metal',
       'Slate', 'Satin Black', 'Macchiato Beige/Black', 'Charcoal',
       'Black / Express Red', 'Cappuccino', 'Aragon Brown', 'Parchment',
       'Oyster W/Contrast', 'Adrenaline Red', 'Ebony.', 'Shara Beige',
       'Graystone', 'Pearl Beige', 'Nero Ade', 'Graphite',
       'Tan/Ebony/Ebony', 'Charcoal Black', 'Medium Ash Gray',
       'Ebony Black', 'Light Titanium', 'Sakhir Orange', 'Tan',
       'Rock Gray', 'Brandy', 'Carbon Black', 'Amber',
      

In [ ]:
def simplify_colors(color):
    color = str(color).lower()
    if 'red' in color:
        return 'red'
    elif 'blue' in color:
        return 'blue'
    elif 'black' in color:
        return 'black'
    elif 'white' in color:
        return 'white'
    else:
        return 'others'

dataset['int_col'] = dataset['int_col'].apply(simplify_colors)
dataset['ext_col'] = dataset['ext_col'].apply(simplify_colors)

In [ ]:
def data_prep(dataset):
  # Chaning columns datatypes and values
  dataset['age'] = (2024 - dataset["model_year"]) + 1

  # Drop Unneccesary Columns
  dataset.drop(['id'] , inplace = True , axis = 1)
  dataset.drop(['model_year'] , inplace = True , axis = 1)
  dataset.drop(['engine'], inplace = True , axis = 1)
  dataset.drop(["model"], inplace = True , axis = 1)

  # Impute Missing values
  dataset['fuel_type'] = np.where(dataset['fuel_type'].isna() ,
                                  dataset['fuel_type'].mode(dropna = True)[0] , dataset['fuel_type'])
  dataset['accident'] = np.where(dataset['accident'].isna() ,
                                  dataset['accident'].mode(dropna = True)[0] , dataset['accident'])
  dataset['fuel_type'] = np.where(dataset['fuel_type'] == '-' ,
                                  dataset['fuel_type'].mode(dropna = True)[0] , dataset['fuel_type'])
  dataset["clean_title"].fillna('No',inplace=True)
  dataset['Horsepower'].fillna(dataset['Horsepower'].mean(), inplace = True)
  dataset['Engine_Displacement'].fillna(dataset['Engine_Displacement'].mean(), inplace = True)


In [ ]:
data_prep(dataset)

In [ ]:
dataset.isna().sum()

brand                  0
milage                 0
fuel_type              0
transmission           0
ext_col                0
int_col                0
accident               0
clean_title            0
price                  0
Horsepower             0
Engine_Displacement    0
age                    0
dtype: int64

In [ ]:
dataset.head()

brand  milage      fuel_type transmission ext_col int_col  \
0           MINI  213000       Gasoline    Automatic  others  others   
1        Lincoln  143250       Gasoline    Automatic  others  others   
2      Chevrolet  136731  E85 Flex Fuel    Automatic    blue  others   
3        Genesis   19500       Gasoline    Automatic   black   black   
4  Mercedes-Benz    7388       Gasoline    Automatic   black  others   

                                 accident clean_title  price  Horsepower  \
0                           None reported         Yes   4200       172.0   
1  At least 1 accident or damage reported         Yes   4999       252.0   
2                           None reported         Yes  13900       320.0   
3                           None reported         Yes  45000       420.0   
4                           None reported         Yes  97500       208.0   

   Engine_Displacement  age  
0                  1.6   18  
1                  3.9   23  
2                  5.3   23  
3                  5.0    8  
4                  2.0    4

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188533 entries, 0 to 188532
Data columns (total 12 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   brand                188533 non-null  object 
 1   milage               188533 non-null  int64  
 2   fuel_type            188533 non-null  object 
 3   transmission         188533 non-null  object 
 4   ext_col              188533 non-null  object 
 5   int_col              188533 non-null  object 
 6   accident             188533 non-null  object 
 7   clean_title          188533 non-null  object 
 8   price                188533 non-null  int64  
 9   Horsepower           188533 non-null  float64
 10  Engine_Displacement  188533 non-null  float64
 11  age                  188533 non-null  int64  
dtypes: float64(2), int64(3), object(7)
memory usage: 17.3+ MB


# Label Encoder

In [ ]:
def data_encode(data):
  le =LabelEncoder()

  for column in data.columns:
      if data[column].dtype=='object':
          data[column] = le.fit_transform(data[column])


In [ ]:
data_encode(dataset)

In [ ]:
dataset.head()

brand  milage  fuel_type  transmission  ext_col  int_col  accident  \
0     31  213000          2             0        2        2         1   
1     28  143250          2             0        2        2         0   
2      9  136731          1             0        1        2         1   
3     16   19500          2             0        0        0         1   
4     36    7388          2             0        0        2         1   

   clean_title  price  Horsepower  Engine_Displacement  age  
0            1   4200       172.0                  1.6   18  
1            1   4999       252.0                  3.9   23  
2            1  13900       320.0                  5.3   23  
3            1  45000       420.0                  5.0    8  
4            1  97500       208.0                  2.0    4

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188533 entries, 0 to 188532
Data columns (total 12 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   brand                188533 non-null  int64  
 1   milage               188533 non-null  int64  
 2   fuel_type            188533 non-null  int64  
 3   transmission         188533 non-null  int64  
 4   ext_col              188533 non-null  int64  
 5   int_col              188533 non-null  int64  
 6   accident             188533 non-null  int64  
 7   clean_title          188533 non-null  int64  
 8   price                188533 non-null  int64  
 9   Horsepower           188533 non-null  float64
 10  Engine_Displacement  188533 non-null  float64
 11  age                  188533 non-null  int64  
dtypes: float64(2), int64(10)
memory usage: 17.3 MB


# Spilt The Dataset

In [ ]:
x = dataset.drop(['price'] , axis = 1)
y = dataset['price']

In [ ]:
x_train , x_test , y_train , y_test = train_test_split(x , y , test_size = 0.2 , random_state = 123)

In [ ]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(150826, 11)
(37707, 11)
(150826,)
(37707,)


# Model Evaluation Function

In [ ]:
def train_and_evaluate(model, X_train, y_train, X_test, y_test):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    rmse = mean_squared_error(y_test, y_pred, squared=False)
    mse = mean_squared_error(y_test, y_pred)
    return model , round(r2,2) , round(rmse,2) , int(round(mse,2))

# Machine Learning Models

In [ ]:
models = {
    'Linear Regression': LinearRegression(),
    'Decision Tree': DecisionTreeRegressor(),
    'Random Forest': RandomForestRegressor(),
    'Ada Boosting': AdaBoostRegressor(),
    'Gradient Boosting': GradientBoostingRegressor(),
    'XGBoosting': XGBRegressor(),
    'KNN': KNeighborsRegressor()
}

In [ ]:
def base_models(models):
  results = []
  for model_name, Model in models.items():
    model , r2 , rmse , mse = train_and_evaluate(Model , x_train , y_train , x_test , y_test)
    results.append([model_name , r2 , rmse , mse])
  return results

df_results = pd.DataFrame(base_models(models), columns=['Model', 'R2 Score', 'RMSE', 'MSE'])
print(df_results)

               Model  R2 Score       RMSE          MSE
0  Linear Regression      0.10   74097.40   5490425367
1      Decision Tree     -0.84  105619.98  11155579667
2      Random Forest     -0.00   77992.07   6082763285
3       Ada Boosting     -2.14  138051.91  19058329334
4  Gradient Boosting      0.14   72234.62   5217839720
5         XGBoosting      0.10   73752.65   5439453640
6                KNN     -0.02   78772.18   6205056488


# Optuna - Tuned Models

In [ ]:
def objective(trial, model, X_train, y_train, X_test, y_test):
    params = {}
    if isinstance(model, RandomForestRegressor):
        params["n_estimators"] = trial.suggest_int("n_estimators", 50, 200)
        params["max_depth"] = trial.suggest_int("max_depth", 10, 100)
        params["min_samples_split"] = trial.suggest_int("min_samples_split", 2, 10)
    elif isinstance(model, GradientBoostingRegressor):
        params["n_estimators"] = trial.suggest_int("n_estimators", 50, 200)
        params["learning_rate"] = trial.suggest_float("learning_rate", 0.01, 0.3)
        params["max_depth"] = trial.suggest_int("max_depth", 3, 10)
    elif isinstance(model, DecisionTreeRegressor):
        params["max_depth"] = trial.suggest_int("max_depth", 3, 10)
        params["min_samples_split"] = trial.suggest_int("min_samples_split", 2, 10)
    elif isinstance(model, AdaBoostRegressor):
        params["n_estimators"] = trial.suggest_int("n_estimators", 50, 200)
        params["learning_rate"] = trial.suggest_float("learning_rate", 0.01, 0.3)
    elif isinstance(model, XGBRegressor):
        params["n_estimators"] = trial.suggest_int("n_estimators", 50, 200)
        params["learning_rate"] = trial.suggest_float("learning_rate", 0.01, 0.3)
        params["max_depth"] = trial.suggest_int("max_depth", 3, 10)
    elif isinstance(model, KNeighborsRegressor):
        params["n_neighbors"] = trial.suggest_int("n_neighbors", 3, 15)
    else:
        params = {}  # For other models, no hyperparameters to tune

    model.set_params(**params)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    return mean_squared_error(y_test, y_pred, squared=False)

In [ ]:
def tune_models(models, X_train, y_train, X_test, y_test, patience=3):
    tuned_models = {}
    for model_name, model in models.items():
        pruner = optuna.pruners.PatientPruner(optuna.pruners.MedianPruner(), patience=patience)
        study = optuna.create_study(direction="minimize", pruner=pruner)
        study.optimize(lambda trial: objective(trial, model, X_train, y_train, X_test, y_test), n_trials=10)

        best_params = study.best_params
        best_model = model.set_params(**best_params)
        best_model.fit(X_train, y_train)
        y_pred = best_model.predict(X_test)
        r2 = r2_score(y_test, y_pred)
        rmse = mean_squared_error(y_test, y_pred, squared=False)
        tuned_models[model_name] = {'model': best_model, 'r2': r2, 'rmse': rmse, 'params': best_params}
        print(f"Best parameters for {model_name}: {best_params}")
    return tuned_models

tuned_models = tune_models(models, x_train, y_train, x_test, y_test)

[I 2024-11-18 12:41:59,866] A new study created in memory with name: no-name-f3df5e6d-77b7-4b0f-9193-f5fdbe039529
[I 2024-11-18 12:41:59,919] Trial 0 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,056] Trial 1 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,123] Trial 2 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,256] Trial 3 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,347] Trial 4 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,452] Trial 5 finished with value: 74097.40459454338 and parameters: {}. Best is trial 0 with value: 74097.40459454338.
[I 2024-11-18 12:42:00,540] Trial 

Best parameters for Linear Regression: {}


[I 2024-11-18 12:42:01,184] Trial 0 finished with value: 76053.42725641282 and parameters: {'max_depth': 6, 'min_samples_split': 7}. Best is trial 0 with value: 76053.42725641282.
[I 2024-11-18 12:42:01,385] Trial 1 finished with value: 73166.43286078265 and parameters: {'max_depth': 4, 'min_samples_split': 8}. Best is trial 1 with value: 73166.43286078265.
[I 2024-11-18 12:42:01,751] Trial 2 finished with value: 76406.33665073976 and parameters: {'max_depth': 7, 'min_samples_split': 4}. Best is trial 1 with value: 73166.43286078265.
[I 2024-11-18 12:42:01,967] Trial 3 finished with value: 73166.43286078265 and parameters: {'max_depth': 4, 'min_samples_split': 5}. Best is trial 1 with value: 73166.43286078265.
[I 2024-11-18 12:42:02,121] Trial 4 finished with value: 73652.41411247413 and parameters: {'max_depth': 3, 'min_samples_split': 10}. Best is trial 1 with value: 73166.43286078265.
[I 2024-11-18 12:42:02,494] Trial 5 finished with value: 78019.27838688373 and parameters: {'max_de

Best parameters for Decision Tree: {'max_depth': 4, 'min_samples_split': 8}


[I 2024-11-18 12:44:35,420] Trial 0 finished with value: 77707.543148889 and parameters: {'n_estimators': 173, 'max_depth': 66, 'min_samples_split': 2}. Best is trial 0 with value: 77707.543148889.
[I 2024-11-18 12:46:47,028] Trial 1 finished with value: 75214.35530183002 and parameters: {'n_estimators': 185, 'max_depth': 80, 'min_samples_split': 7}. Best is trial 1 with value: 75214.35530183002.
[I 2024-11-18 12:49:01,482] Trial 2 finished with value: 75594.09477942137 and parameters: {'n_estimators': 183, 'max_depth': 29, 'min_samples_split': 5}. Best is trial 1 with value: 75214.35530183002.
[I 2024-11-18 12:50:19,012] Trial 3 finished with value: 75424.57965279977 and parameters: {'n_estimators': 110, 'max_depth': 30, 'min_samples_split': 7}. Best is trial 1 with value: 75214.35530183002.
[I 2024-11-18 12:51:51,552] Trial 4 finished with value: 74334.50244519032 and parameters: {'n_estimators': 138, 'max_depth': 95, 'min_samples_split': 10}. Best is trial 4 with value: 74334.502445

Best parameters for Random Forest: {'n_estimators': 138, 'max_depth': 95, 'min_samples_split': 10}


[I 2024-11-18 13:02:22,314] Trial 0 finished with value: 94857.10809162239 and parameters: {'n_estimators': 171, 'learning_rate': 0.10074959366367599}. Best is trial 0 with value: 94857.10809162239.
[I 2024-11-18 13:02:33,486] Trial 1 finished with value: 93479.01263909467 and parameters: {'n_estimators': 83, 'learning_rate': 0.11608823585254549}. Best is trial 1 with value: 93479.01263909467.
[I 2024-11-18 13:03:00,340] Trial 2 finished with value: 80033.90254252877 and parameters: {'n_estimators': 118, 'learning_rate': 0.021902288519762628}. Best is trial 2 with value: 80033.90254252877.
[I 2024-11-18 13:03:12,192] Trial 3 finished with value: 85825.13415832717 and parameters: {'n_estimators': 58, 'learning_rate': 0.11355046788254494}. Best is trial 2 with value: 80033.90254252877.
[I 2024-11-18 13:03:18,537] Trial 4 finished with value: 85289.08152553652 and parameters: {'n_estimators': 57, 'learning_rate': 0.2853457330683366}. Best is trial 2 with value: 80033.90254252877.
[I 2024-

Best parameters for Ada Boosting: {'n_estimators': 118, 'learning_rate': 0.021902288519762628}


[I 2024-11-18 13:04:48,821] Trial 0 finished with value: 83416.6973852087 and parameters: {'n_estimators': 59, 'learning_rate': 0.27746635370402273, 'max_depth': 10}. Best is trial 0 with value: 83416.6973852087.
[I 2024-11-18 13:05:48,737] Trial 1 finished with value: 81504.4942816097 and parameters: {'n_estimators': 127, 'learning_rate': 0.1745408498990973, 'max_depth': 9}. Best is trial 1 with value: 81504.4942816097.
[I 2024-11-18 13:06:13,620] Trial 2 finished with value: 72320.91333705166 and parameters: {'n_estimators': 153, 'learning_rate': 0.051895998160992315, 'max_depth': 3}. Best is trial 2 with value: 72320.91333705166.
[I 2024-11-18 13:06:49,253] Trial 3 finished with value: 75565.8473745639 and parameters: {'n_estimators': 137, 'learning_rate': 0.25490052749568265, 'max_depth': 5}. Best is trial 2 with value: 72320.91333705166.
[I 2024-11-18 13:07:24,486] Trial 4 finished with value: 76224.18219952857 and parameters: {'n_estimators': 68, 'learning_rate': 0.03080933192658

Best parameters for Gradient Boosting: {'n_estimators': 153, 'learning_rate': 0.051895998160992315, 'max_depth': 3}


[I 2024-11-18 13:10:30,031] Trial 0 finished with value: 77856.21175549616 and parameters: {'n_estimators': 187, 'learning_rate': 0.17908661939802423, 'max_depth': 9}. Best is trial 0 with value: 77856.21175549616.
[I 2024-11-18 13:10:31,279] Trial 1 finished with value: 72864.80180762328 and parameters: {'n_estimators': 150, 'learning_rate': 0.20580208597737598, 'max_depth': 5}. Best is trial 1 with value: 72864.80180762328.
[I 2024-11-18 13:10:33,266] Trial 2 finished with value: 75759.87012908435 and parameters: {'n_estimators': 122, 'learning_rate': 0.11381717705380015, 'max_depth': 9}. Best is trial 1 with value: 72864.80180762328.
[I 2024-11-18 13:10:34,404] Trial 3 finished with value: 73082.45683345241 and parameters: {'n_estimators': 110, 'learning_rate': 0.14490644440253217, 'max_depth': 6}. Best is trial 1 with value: 72864.80180762328.
[I 2024-11-18 13:10:36,772] Trial 4 finished with value: 79216.79896286668 and parameters: {'n_estimators': 120, 'learning_rate': 0.24101407

Best parameters for XGBoosting: {'n_estimators': 142, 'learning_rate': 0.05513954840177207, 'max_depth': 4}


[I 2024-11-18 13:10:46,688] Trial 0 finished with value: 76473.67809045728 and parameters: {'n_neighbors': 8}. Best is trial 0 with value: 76473.67809045728.
[I 2024-11-18 13:10:47,478] Trial 1 finished with value: 74972.54740150577 and parameters: {'n_neighbors': 11}. Best is trial 1 with value: 74972.54740150577.
[I 2024-11-18 13:10:48,354] Trial 2 finished with value: 74558.70378096418 and parameters: {'n_neighbors': 14}. Best is trial 2 with value: 74558.70378096418.
[I 2024-11-18 13:10:49,173] Trial 3 finished with value: 77214.87332735915 and parameters: {'n_neighbors': 7}. Best is trial 2 with value: 74558.70378096418.
[I 2024-11-18 13:10:50,074] Trial 4 finished with value: 77214.87332735915 and parameters: {'n_neighbors': 7}. Best is trial 2 with value: 74558.70378096418.
[I 2024-11-18 13:10:51,561] Trial 5 finished with value: 74558.70378096418 and parameters: {'n_neighbors': 14}. Best is trial 2 with value: 74558.70378096418.
[I 2024-11-18 13:10:53,062] Trial 6 finished with

Best parameters for KNN: {'n_neighbors': 15}


In [ ]:
# Create a list to store the results
results = []

for model_name, model_data in tuned_models.items():
    results.append([
        model_name,
        model_data['r2'],
        model_data['rmse'],
        mean_squared_error(y_test, model_data['model'].predict(x_test)),  # Calculate MSE
        model_data['params']
    ])

# Create the DataFrame
df_optimized_models = pd.DataFrame(results, columns=['Model', 'R2 Score', 'RMSE', 'MSE', 'Hyperparameters'])

# Display the DataFrame
print(df_optimized_models)

# Find the best model based on R2 score (you can change this to RMSE or another metric)
best_model_name = df_optimized_models.loc[df_optimized_models['RMSE'].idxmin(), 'Model']
print(f"\nBest Model (based on low RMSE value): {best_model_name}")

               Model  R2 Score          RMSE           MSE  \
0  Linear Regression  0.096464  74097.404595  5.490425e+09   
1      Decision Tree  0.119025  73166.432861  5.353327e+09   
2      Random Forest  0.083057  74645.107890  5.571892e+09   
3       Ada Boosting -0.074609  80808.188022  6.529963e+09   
4  Gradient Boosting  0.139269  72320.913337  5.230315e+09   
5         XGBoosting  0.143496  72143.090919  5.204626e+09   
6                KNN  0.086454  74506.723818  5.551252e+09   

                                     Hyperparameters  
0                                                 {}  
1           {'max_depth': 4, 'min_samples_split': 8}  
2  {'n_estimators': 138, 'max_depth': 95, 'min_sa...  
3  {'n_estimators': 118, 'learning_rate': 0.02190...  
4  {'n_estimators': 153, 'learning_rate': 0.05189...  
5  {'n_estimators': 142, 'learning_rate': 0.05513...  
6                                {'n_neighbors': 15}  

Best Model (based on low RMSE value): XGBoosting


# Deployment - Gradio App

In [ ]:
def predict_price(brand ,milage ,fuel_type, transmission, ext_col, int_col, accident,clean_title, horsepower, engine_displacement,age ):
  # Create a DataFrame from the input features
  input_data = pd.DataFrame({
      'brand': [brand],
      'milage': [milage],
      'fuel_type': [fuel_type],
      'transmission': [transmission],
      'ext_col': [ext_col],
      'int_col': [int_col],
      'accident': [accident],
      'clean_title': [clean_title],
      'Horsepower': [horsepower],
      'Engine_Displacement': [engine_displacement],
      'age': [age],

  })

  # Preprocess the input data (similar to how you preprocessed the training data)
  data_encode(input_data) # Assuming data_encode function is defined as in your code

  # Make prediction using the best tuned model
  best_model = tuned_models[best_model_name]['model']
  predicted_price = best_model.predict(input_data)[0]

  return predicted_price


iface = gr.Interface(
    fn=predict_price,
    inputs=[
        gr.Dropdown(choices=['MINI', 'Lincoln', 'Chevrolet', 'Genesis', 'Mercedes-Benz', 'Audi',
                             'Ford', 'BMW', 'Tesla', 'Cadillac', 'Land', 'GMC', 'Toyota',
                             'Hyundai', 'Volvo', 'Volkswagen', 'Buick', 'Rivian', 'RAM',
                             'Hummer', 'Alfa', 'INFINITI', 'Jeep', 'Porsche', 'McLaren',
                             'Honda', 'Lexus', 'Dodge', 'Nissan', 'Jaguar', 'Acura', 'Kia',
                             'Mitsubishi', 'Rolls-Royce', 'Maserati', 'Pontiac', 'Saturn',
                             'Bentley', 'Mazda', 'Subaru', 'Ferrari', 'Aston', 'Lamborghini',
                             'Chrysler', 'Lucid', 'Lotus', 'Scion', 'smart', 'Karma',
                             'Plymouth', 'Suzuki', 'FIAT', 'Saab', 'Bugatti', 'Mercury','Polestar', 'Maybach'] , label ="Brand"),
        gr.Number(label="Milage"),
        gr.Dropdown(choices=['Gasoline','E85 Flex Fuel','Hybrid','Diesel','Plug-In Hybrid','Not supported'], label="Fuel Type"),

        gr.Dropdown(choices=['Automatic', 'Manual', 'Other'], label="Transmission"),
        gr.Dropdown(choices=['Red','Blue','Black','White','Others'], label="Exterior Color"),
        gr.Dropdown(choices=['Red','Blue','Black','White','Others'], label="Interior Color"),
        gr.Dropdown(choices=['None reported', 'At least 1 accident or damage reported'], label="Accident"),
        gr.Dropdown(choices=['Yes' , 'No'], label="Clean Title"),
        gr.Number(label="Horsepower"),
        gr.Number(label="Engine Displacement"),
        gr.Number(label="Age of the Car")
    ],
    outputs=gr.Number(label="Predicted Price in $"),
    title="Used Car Price Prediction",
    description="Predict the price of a used car based on various features.",
)

iface.launch()

Running Gradio in a Colab notebook requires sharing enabled. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://898f493f30a9136ae7.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)
